### ISSUE TITLE
Bulk selection for patch level and risk of breaking changes

### ISSUE THREAD

**Issue by deniciocode** [mattbrictson/bundle_update_interactive#19]
### Use Case
As a **user** I would like to **select all gems** 
based on the level of risk it takes to update them. 

### For example: 
If I only want to update gems which will not break. 
In this case a keymap to select all of them would be very helpful.

**Comment by mattbrictson** [comment]
Interesting idea! Do you have a suggested key mapping for this? Right now our key mappings look like:

- **j/k/n/p** vi and emacs up/down
- **a/r** select all and reverse select (see #16)

Perhaps this?

- **shift-p** select [P]atches
- **shift-i** select m[I]nor
- **shift-j** select ma[J]or
- **shift-s** select [S]ecurity

**Comment by deniciocode** [comment]
Hey @mattbrictson 
I do not have an idea what could be the mapping here. Your suggestion sound valid to me.

**Comment by mattbrictson** [comment]
@sirwolfgang I agree; I think there are two slightly different features being discussed.

Right now, `update-interactive` resolves the Gemfile by trying to find the latest upgradeable version for each gem, and then displays those in a table for you to pick and choose. We can offer shortcuts to choose a subset of these updates based on risk. I think this is the feature originally requested in the issue.

A different approach would be to limit the Gemfile resolution algorithm itself, to only allow certain updates, similar to running `bundle update --patch`.

For example, consider your lock file currently has rack 2.2.7. To limit risk, you might want to update to rack 2.2.9 (the latest in the 2.x series) before going all the way to the latest, which is rack 3.1.7. Right now `bundle-update` will only give you the 3.1.7 option.

Because of the complexity of Bundler's dependency resolution algorithm, `bundle-interactive` can't know that both the 2.2.9 and the 3.1.7 options exist without having to re-resolve the Gemfile. So I think this is an option that needs to be provided up front, like `bundle update-interactive --patch`. I'll open a separate issue to capture this as a feature request.

**Comment by brandondrew** [comment]
@mattbrictson How about:

| keystroke† | purpose |
|--------------|--------|
| <kbd>option</kbd>-<kbd>p</kbd> | select patch-level updates|
| <kbd>option</kbd>-<kbd>m</kbd> | select minor updates|
| <kbd>option</kbd>-<kbd>shift</kbd>-<kbd>m</kbd> | select MAJOR updates |
| <kbd>option</kbd>-<kbd>s</kbd> | select security updates |

† <kbd>option</kbd> would be <kbd>alt</kbd> on PC keyboards

**Comment by mattbrictson** [comment]
@brandondrew I don't have a strong preference on the key bindings; your suggestions seem reasonable to me. Although, I am not sure how easy/reliable it is to detect whether a keyboard uses "alt" or "option". To show a good help message, it would be nice to display the key binding appropriate to the user's keyboard.

BTW, I currently don't have plans to implement this enhancement. I would be happy to accept a PR, though.

**Comment by brandondrew** [comment]
The simplest method would be to use the user's OS:

macOS: ⌥ Option
Linux:   Alt
Windows: Alt

Although a Mac user with a Logitech keyboard (for example) might actually have "alt" on the key (or both "alt" and "option"), the OS itself will refer to the key as "option" or with the symbol "⌥".   Some older Apple keyboards had both "option" and "alt", presumably to make it easier for switchers from Windows—since that was a huge market in the early days of Mac OS X.

A Linux user might occasionally be using Apple hardware, but again their OS will generally refer to the key as "alt", so I think it's not a problem to make assumptions based on the OS.

It might be _possible_ to get access to the actual keyboard information in Ruby—I don't know—but it seems like potentially a lot of extra work for a very small pay off.  A dubious payoff, in fact, since (for example) a Mac user with a Dell keyboard probably still thinks of the key as "option" even if the keyboard says "alt".  Simply using the OS to decide what to call the key is easiest and also gives the user consistent UX.
